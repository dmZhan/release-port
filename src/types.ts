export interface execResult {
  out: string
  err: string
  error?: Error
}
