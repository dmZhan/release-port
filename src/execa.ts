import { spawn, spawnSync } from 'node:child_process'
import process from 'node:process'
import { Buffer } from 'node:buffer'
import type { execResult } from './types'

export function execa(otherArgs: string): Promise<execResult> {
  try {
    if (otherArgs === '') {
      return Promise.resolve({
        out: 'Execa nothing',
        err: '',
      })
    }

    const { cmd, arg } = getCommandByPlatform()

    const child = spawn(cmd, [arg, otherArgs], { stdio: 'pipe', cwd: process.cwd() })

    return new Promise((resolve, _reject) => {
      const stdout: string[] = []
      const stderr: string[] = []

      if (child.stdout) {
        child.stdout.on('data', (data) => {
          if (Buffer.isBuffer(data)) {
            return stdout.push(data.toString())
          }
          stdout.push(data)
        })
      }

      if (child.stderr) {
        child.stderr.on('data', (data) => {
          if (Buffer.isBuffer(data)) {
            return stderr.push(data.toString())
          }
          stderr.push(JSON.stringify(data))
        })
      }

      const getDefaultResult = () => {
        const err = stderr.join('\n')
        const out = stdout.join('\n')

        return { err, out }
      }

      child.on('error', error => resolve({ ...getDefaultResult(), error }))
      child.on('close', _code => resolve({ ...getDefaultResult(), error: undefined }))
    })
  }
  catch (error) {
    return Promise.reject(error)
  }
}

export function execaSync(otherArgs: string): execResult {
  const { cmd, arg } = getCommandByPlatform()

  if (otherArgs === '') {
    return {
      out: 'Execasync nothing',
      err: '',
    }
  }

  const result = spawnSync(cmd, [arg, otherArgs], { cwd: process.cwd(), encoding: 'utf-8' })

  return {
    out: Buffer.isBuffer(result.stdout) ? result.stdout.toString() : result.stdout,
    err: Buffer.isBuffer(result.stderr) ? result.stderr.toString() : result.stderr,
    error: result.error,
  }
}

function getCommandByPlatform(): { cmd: 'cmd', arg: '/C' } | { cmd: 'sh', arg: '-c' } {
  const { platform } = process

  if (platform === 'win32') {
    return { cmd: 'cmd', arg: '/C' }
  }
  else {
    return { cmd: 'sh', arg: '-c' }
  }
}
