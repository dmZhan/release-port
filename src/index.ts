export * from './releasePort'
export * from './parseArgs'
export * from './execa'
