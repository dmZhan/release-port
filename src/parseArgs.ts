import process from 'node:process'

function splitArgObjects(args: string[]) {
  const newArgs = []
  let index = 0

  while (index < args.length) {
    const arg = args[index]

    if (arg.includes('{')) {
      const temp = []

      while (!args[index].includes('}')) {
        temp.push(args[index])
        index++
      }
      temp.push(args[index])

      newArgs.push(temp.join(' ').replace(/([\w-]+):\s*([\w-]*)/g, '"$1": "$2"'))
    }
    else {
      newArgs.push(arg)
    }

    index++
  }

  return newArgs
}

export function parse(args: string[] = []) {
  if (!args.length) {
    args = process.argv.slice(2)
  }

  if (args[0] && args[0].match(/node$/)) {
    args = args.slice(2)
  }

  const newArgs = splitArgObjects(args)
  function parseArgs(args: string[], obj: { unknown: string[], [k: string]: any }) {
    // Check if end reached
    if (!args.length) {
      return obj
    }

    const arg = args[0]

    // if statement match conditions:
    // 1. --key=value || -key=value
    // 2. --no-key
    // 3. --key value|nothing
    // else add to unknown arr
    if (/^-.+=/.test(arg)) {
      const match = arg.match(/^(--|-)([^=]+)=([\s\S]*)$/) as string[]
      // Set key(match[2]) = value(match[3])
      obj[match[2]] = match[3]
    }
    else if (/^--no-.+/.test(arg)) {
      // Set key = true
      obj[arg.match(/^--no-(.+)/)![1]] = false
    }
    else if (/^-.+/.test(arg)) {
      const key = arg.match(/^(--|-)(.+)/)![2]
      const next = args[1]

      // If next value exist and not prefixed with - or --
      if (next && !/^-{1,2}/.test(next)) {
        obj[key] = next
        return parseArgs(args.slice(2), obj)
      }
      else {
        obj[key] = true
      }
    }
    else {
      obj.unknown.push(arg)
    }

    return parseArgs(args.slice(1), obj)
  }

  const parseResult = parseArgs(newArgs, { unknown: [] })

  // Covert to proper type
  for (const prop in parseResult) {
    try {
      parseResult[prop] = JSON.parse(parseResult[prop])
    }
    catch (e) {
      continue
    }
  }

  return parseResult
}
