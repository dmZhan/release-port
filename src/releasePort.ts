import process from 'node:process'
import { execa, execaSync } from './execa'

/**
 * kill port
 * @param port port number
 * @param method 'tcp' | 'udp'
 * @returns Promise
 */
export async function release(port: number | string, method: 'tcp' | 'udp' = 'tcp') {
  const p = typeof port === 'string' ? Number.parseInt(port) : port

  if (!p) {
    return Promise.reject(new Error('Invalid port number provided'))
  }

  if (process.platform === 'win32') {
    return execa('netstat -nao').then((res) => {
      const { out } = res

      if (!out) {
        return res
      }

      return execa(killPort('win32', getPids(out, method, p)))
    })
  }
  else {
    return execa('lsof -i -p').then((res) => {
      const { out } = res

      if (!out) {
        return res
      }

      if (!isExistProcess(out, p)) {
        return Promise.reject(new Error('No process running on port'))
      }

      return execa(
        killPort(process.platform as Exclude<NodeJS.Platform, 'win32'>, method, p),
      )
    })
  }
}

/**
 * sync kill port
 * @param port port number
 * @param method 'tcp' | 'udp'
 * @returns execResult | Error
 */
export function releaseSync(port: number | string, method: 'tcp' | 'udp' = 'tcp') {
  const p = typeof port === 'string' ? Number.parseInt(port) : port

  if (!p) {
    return new Error('Invalid port number provided')
  }

  if (process.platform === 'win32') {
    const res = execaSync('netstat -nao')

    if (!res.out) {
      return res
    }

    return execaSync(killPort('win32', getPids(res.out, method, p)))
  }
  else {
    const res = execaSync('lsof -i -p')

    if (!res.out) {
      return res
    }

    if (!isExistProcess(res.out, p)) {
      return new Error('No process running on port')
    }

    return execaSync(
      killPort(process.platform as Exclude<NodeJS.Platform, 'win32'>, method, p),
    )
  }
}

function killPort(platform: 'win32', pids: string[]): string
function killPort(platform: Exclude<NodeJS.Platform, 'win32'>, method: 'udp' | 'tcp', port: number): string
function killPort(platform: NodeJS.Platform, ops?: any, port?: number): string {
  if (platform === 'win32') {
    return `TaskKill /F /PID ${ops?.join(' /PID ')}`
  }
  else {
    return `lsof -i ${ops === 'udp' ? 'udp' : 'tcp'}:${port} | grep ${ops === 'udp' ? 'UDP' : 'LISTEN'} | awk '{print $2}' | xrags kill -9`
  }
}

/**
 * get pids in windows
 * @param str pids output string
 * @param method 'udp' | 'tcp'
 * @param port port number
 * @returns string[]
 */
function getPids(str: string, method: 'udp' | 'tcp', port: number) {
  const lines = str.split('\n')
  const lineWithLocalPortRegEx = new RegExp(`^ *${method.toUpperCase()} *[^ ]*:${port}`, 'gm')
  const linesWithLocalPort = lines.filter(line => line.match(lineWithLocalPortRegEx))

  const pids = linesWithLocalPort.reduce((acc: string[], line) => {
    const match = line.match(/(\d*)(\n|$)/gm)
    return match && match[0] && !acc.includes(match[0]) ? acc.concat(match[0]) : acc
  }, [])

  return pids
}

/**
 * process whether exist
 * @param str pids output string
 * @param port port number
 * @returns boolean
 */
function isExistProcess(str: string, port: number) {
  const lines = str.split('\n')
  return lines.filter(line => line.match(new RegExp(`:*${port}`))).length > 0
}
