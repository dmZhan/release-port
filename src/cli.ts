import process from 'node:process'
import { release } from './releasePort'
import { parse } from './parseArgs'

/* eslint-disable no-console */
function cli() {
  const args = parse(process.argv.slice(2))
  const verbose = args.verbose || false

  let port = args.port ? args.port.toString().split(',') : args.unknown
  const method = args.method || 'tcp'

  if (!Array.isArray(port)) {
    port = [port]
  }

  Promise.all(port.map(async (current: string | number) => {
    return await release(current, method)
      .then((result) => {
        console.log(`Process on port ${current} killed`)
        verbose && console.log(result)
      })
      .catch((error) => {
        console.log(`Could not release process on port ${port}. ${error.message}.`)
        verbose && console.log(error)
      })
  }))
}

cli()
