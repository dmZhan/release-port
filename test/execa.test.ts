import { describe, expect, it } from 'vitest'
import { execa, execaSync } from '../src/execa'

describe('execa', () => {
  it('execaSync', () => {
    expect(execaSync('echo 123')).toMatchInlineSnapshot(`
      {
        "err": "",
        "error": undefined,
        "out": "123
      ",
      }
    `)

    expect(execaSync('echoa 123')).toMatchInlineSnapshot(`
      {
        "err": "'echoa' is not recognized as an internal or external command,
      operable program or batch file.
      ",
        "error": undefined,
        "out": "",
      }
    `)

    expect(execaSync('echo 123')).toEqual({
      out: expect.anything(),
      err: expect.anything(),
      error: undefined,
    })
  })

  it('execa', async () => {
    expect(await execa('echo 123')).toMatchInlineSnapshot(`
      {
        "err": "",
        "error": undefined,
        "out": "123
      ",
      }
    `)

    expect(await execa('echoa 123')).toMatchInlineSnapshot(`
      {
        "err": "'echoa' is not recognized as an internal or external command,
      operable program or batch file.
      ",
        "error": undefined,
        "out": "",
      }
    `)

    expect(await execa('echo 123')).toEqual({
      out: expect.anything(),
      err: expect.anything(),
      error: undefined,
    })
  })

  it('same command\'s output should equal', async () => {
    expect(await execa('echo 123')).toEqual(execaSync('echo 123'))

    expect(execaSync('echoa 123')).toEqual(await execa('echoa 123'))
  })

  it.skip('test ommand netstat -nao', () => {
    expect(execaSync('netstat -nao')).toMatchSnapshot()
  })
})
