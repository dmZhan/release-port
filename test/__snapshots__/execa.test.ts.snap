// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`execa > test ommand netstat -nao 1`] = `
{
  "err": "",
  "error": undefined,
  "out": "
Active Connections

  Proto  Local Address          Foreign Address        State           PID
  TCP    0.0.0.0:22             0.0.0.0:0              LISTENING       5404
  TCP    0.0.0.0:135            0.0.0.0:0              LISTENING       1160
  TCP    0.0.0.0:445            0.0.0.0:0              LISTENING       4
  TCP    0.0.0.0:2179           0.0.0.0:0              LISTENING       2536
  TCP    0.0.0.0:3306           0.0.0.0:0              LISTENING       5988
  TCP    0.0.0.0:3389           0.0.0.0:0              LISTENING       1512
  TCP    0.0.0.0:5040           0.0.0.0:0              LISTENING       10976
  TCP    0.0.0.0:6881           0.0.0.0:0              LISTENING       26980
  TCP    0.0.0.0:7680           0.0.0.0:0              LISTENING       3544
  TCP    0.0.0.0:10050          0.0.0.0:0              LISTENING       5552
  TCP    0.0.0.0:24678          0.0.0.0:0              LISTENING       10676
  TCP    0.0.0.0:33060          0.0.0.0:0              LISTENING       5988
  TCP    0.0.0.0:41407          0.0.0.0:0              LISTENING       6120
  TCP    0.0.0.0:49664          0.0.0.0:0              LISTENING       748
  TCP    0.0.0.0:49665          0.0.0.0:0              LISTENING       956
  TCP    0.0.0.0:49666          0.0.0.0:0              LISTENING       1964
  TCP    0.0.0.0:49667          0.0.0.0:0              LISTENING       2016
  TCP    0.0.0.0:49668          0.0.0.0:0              LISTENING       3140
  TCP    0.0.0.0:49669          0.0.0.0:0              LISTENING       4572
  TCP    0.0.0.0:49675          0.0.0.0:0              LISTENING       4608
  TCP    0.0.0.0:49700          0.0.0.0:0              LISTENING       92
  TCP    0.0.0.0:50051          0.0.0.0:0              LISTENING       5032
  TCP    0.0.0.0:50091          0.0.0.0:0              LISTENING       5184
  TCP    0.0.0.0:50094          0.0.0.0:0              LISTENING       17196
  TCP    0.0.0.0:50095          0.0.0.0:0              LISTENING       23396
  TCP    10.30.247.227:139      0.0.0.0:0              LISTENING       4
  TCP    10.30.247.227:52007    10.0.0.11:37527        ESTABLISHED     5040
  TCP    10.30.247.227:52222    20.197.71.89:443       ESTABLISHED     5412
  TCP    10.30.247.227:52561    101.91.37.18:80        ESTABLISHED     18092
  TCP    10.30.247.227:54613    104.16.132.229:443     ESTABLISHED     12560
  TCP    10.30.247.227:56241    120.53.53.53:443       ESTABLISHED     12560
  TCP    10.30.247.227:57693    61.172.229.212:21007   ESTABLISHED     12560
  TCP    10.30.247.227:58009    61.172.229.212:21007   TIME_WAIT       0
  TCP    10.30.247.227:58326    223.109.107.98:443     TIME_WAIT       0
  TCP    10.30.247.227:59631    61.172.229.212:21007   TIME_WAIT       0
  TCP    10.30.247.227:59641    61.172.229.212:21007   TIME_WAIT       0
  TCP    10.30.247.227:59671    61.172.229.212:21007   ESTABLISHED     12560
  TCP    10.30.247.227:59684    223.6.6.6:443          ESTABLISHED     12560
  TCP    10.30.247.227:59818    220.181.174.161:443    TIME_WAIT       0
  TCP    10.30.247.227:59868    112.16.227.102:80      TIME_WAIT       0
  TCP    10.30.247.227:59939    10.0.0.11:80           TIME_WAIT       0
  TCP    10.30.247.227:59944    10.0.0.11:80           TIME_WAIT       0
  TCP    10.30.247.227:59949    20.189.173.10:443      TIME_WAIT       0
  TCP    10.30.247.227:59950    8.8.4.4:853            TIME_WAIT       0
  TCP    10.30.247.227:59951    8.8.4.4:853            TIME_WAIT       0
  TCP    10.30.247.227:59957    61.172.229.212:21007   TIME_WAIT       0
  TCP    10.30.247.227:59963    61.172.229.212:21007   TIME_WAIT       0
  TCP    10.30.247.227:59970    61.172.229.212:21007   TIME_WAIT       0
  TCP    10.30.247.227:59974    20.189.173.21:443      TIME_WAIT       0
  TCP    10.30.247.227:59982    61.172.229.212:21007   TIME_WAIT       0
  TCP    10.30.247.227:59988    112.16.227.102:80      TIME_WAIT       0
  TCP    10.30.247.227:59993    8.8.4.4:853            TIME_WAIT       0
  TCP    10.30.247.227:60015    8.8.4.4:853            TIME_WAIT       0
  TCP    10.30.247.227:60037    20.42.73.25:443        TIME_WAIT       0
  TCP    10.30.247.227:60038    8.8.4.4:853            TIME_WAIT       0
  TCP    10.30.247.227:60060    10.0.0.11:80           TIME_WAIT       0
  TCP    10.30.247.227:60067    10.0.0.11:80           TIME_WAIT       0
  TCP    10.30.247.227:60072    104.208.16.88:443      TIME_WAIT       0
  TCP    10.30.247.227:60075    8.8.4.4:853            TIME_WAIT       0
  TCP    10.30.247.227:60086    20.189.173.21:443      TIME_WAIT       0
  TCP    10.30.247.227:60089    104.208.16.88:443      ESTABLISHED     12560
  TCP    10.30.247.227:60103    61.172.229.212:21007   ESTABLISHED     12560
  TCP    10.30.247.227:60109    61.172.229.212:21007   ESTABLISHED     12560
  TCP    10.30.247.227:60111    61.172.229.212:21007   ESTABLISHED     12560
  TCP    10.30.247.227:60126    192.168.0.2:445        SYN_SENT        4
  TCP    10.30.247.227:60128    192.168.0.2:139        SYN_SENT        4
  TCP    10.30.247.227:60138    223.109.107.97:443     ESTABLISHED     12560
  TCP    10.30.247.227:60148    192.168.0.119:50091    SYN_SENT        12560
  TCP    10.30.247.227:60151    61.172.229.212:21007   TIME_WAIT       0
  TCP    10.30.247.227:60153    61.172.229.212:21007   ESTABLISHED     12560
  TCP    10.30.247.227:60155    192.168.0.119:50091    SYN_SENT        12560
  TCP    10.30.247.227:60156    192.168.0.1:20000      SYN_SENT        5032
  TCP    127.0.0.1:6379         0.0.0.0:0              LISTENING       5272
  TCP    127.0.0.1:7890         0.0.0.0:0              LISTENING       12560
  TCP    127.0.0.1:7890         127.0.0.1:53151        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53152        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53153        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53194        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53195        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53245        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53246        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53254        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53263        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53270        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53287        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53329        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53352        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53369        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53378        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53416        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53428        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53447        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:53458        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:57692        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:59586        TIME_WAIT       0
  TCP    127.0.0.1:7890         127.0.0.1:59670        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:60088        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:60102        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:60108        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:60110        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:60136        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:60147        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:60152        ESTABLISHED     12560
  TCP    127.0.0.1:7890         127.0.0.1:60154        ESTABLISHED     12560
  TCP    127.0.0.1:8680         0.0.0.0:0              LISTENING       18092
  TCP    127.0.0.1:8828         0.0.0.0:0              LISTENING       13412
  TCP    127.0.0.1:8829         0.0.0.0:0              LISTENING       30200
  TCP    127.0.0.1:8830         0.0.0.0:0              LISTENING       996
  TCP    127.0.0.1:8831         0.0.0.0:0              LISTENING       21684
  TCP    127.0.0.1:8832         0.0.0.0:0              LISTENING       15972
  TCP    127.0.0.1:27017        0.0.0.0:0              LISTENING       5112
  TCP    127.0.0.1:36600        0.0.0.0:0              LISTENING       5040
  TCP    127.0.0.1:42950        0.0.0.0:0              LISTENING       848
  TCP    127.0.0.1:49684        127.0.0.1:49685        ESTABLISHED     5988
  TCP    127.0.0.1:49685        127.0.0.1:49684        ESTABLISHED     5988
  TCP    127.0.0.1:49686        127.0.0.1:49687        ESTABLISHED     5988
  TCP    127.0.0.1:49687        127.0.0.1:49686        ESTABLISHED     5988
  TCP    127.0.0.1:52464        127.0.0.1:52465        ESTABLISHED     26980
  TCP    127.0.0.1:52465        127.0.0.1:52464        ESTABLISHED     26980
  TCP    127.0.0.1:52612        0.0.0.0:0              LISTENING       15912
  TCP    127.0.0.1:52614        0.0.0.0:0              LISTENING       12560
  TCP    127.0.0.1:53151        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53152        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53153        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53194        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53195        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53245        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53246        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53254        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53263        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53270        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53287        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53329        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53352        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53369        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53378        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53416        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53428        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53447        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:53458        127.0.0.1:7890         ESTABLISHED     6120
  TCP    127.0.0.1:54545        0.0.0.0:0              LISTENING       13412
  TCP    127.0.0.1:54546        0.0.0.0:0              LISTENING       30200
  TCP    127.0.0.1:54554        0.0.0.0:0              LISTENING       21684
  TCP    127.0.0.1:54561        0.0.0.0:0              LISTENING       996
  TCP    127.0.0.1:57692        127.0.0.1:7890         ESTABLISHED     27548
  TCP    127.0.0.1:58008        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:58324        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:59630        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:59640        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:59670        127.0.0.1:7890         ESTABLISHED     27548
  TCP    127.0.0.1:59718        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:59817        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:59945        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:59948        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:59958        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:59973        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:60036        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:60071        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:60085        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:60088        127.0.0.1:7890         ESTABLISHED     13220
  TCP    127.0.0.1:60102        127.0.0.1:7890         ESTABLISHED     27548
  TCP    127.0.0.1:60108        127.0.0.1:7890         ESTABLISHED     27548
  TCP    127.0.0.1:60110        127.0.0.1:7890         ESTABLISHED     27548
  TCP    127.0.0.1:60112        127.0.0.1:7890         TIME_WAIT       0
  TCP    127.0.0.1:60136        127.0.0.1:7890         ESTABLISHED     27548
  TCP    127.0.0.1:60147        127.0.0.1:7890         ESTABLISHED     17196
  TCP    127.0.0.1:60152        127.0.0.1:7890         ESTABLISHED     23024
  TCP    127.0.0.1:60154        127.0.0.1:7890         ESTABLISHED     23396
  TCP    172.27.224.1:139       0.0.0.0:0              LISTENING       4
  TCP    [::]:22                [::]:0                 LISTENING       5404
  TCP    [::]:135               [::]:0                 LISTENING       1160
  TCP    [::]:445               [::]:0                 LISTENING       4
  TCP    [::]:2179              [::]:0                 LISTENING       2536
  TCP    [::]:3306              [::]:0                 LISTENING       5988
  TCP    [::]:3389              [::]:0                 LISTENING       1512
  TCP    [::]:6881              [::]:0                 LISTENING       26980
  TCP    [::]:7680              [::]:0                 LISTENING       3544
  TCP    [::]:10050             [::]:0                 LISTENING       5552
  TCP    [::]:24678             [::]:0                 LISTENING       10676
  TCP    [::]:33060             [::]:0                 LISTENING       5988
  TCP    [::]:41407             [::]:0                 LISTENING       6120
  TCP    [::]:49664             [::]:0                 LISTENING       748
  TCP    [::]:49665             [::]:0                 LISTENING       956
  TCP    [::]:49666             [::]:0                 LISTENING       1964
  TCP    [::]:49667             [::]:0                 LISTENING       2016
  TCP    [::]:49668             [::]:0                 LISTENING       3140
  TCP    [::]:49669             [::]:0                 LISTENING       4572
  TCP    [::]:49675             [::]:0                 LISTENING       4608
  TCP    [::]:49700             [::]:0                 LISTENING       92
  TCP    [::]:50051             [::]:0                 LISTENING       5032
  TCP    [::]:50091             [::]:0                 LISTENING       5184
  TCP    [::]:50094             [::]:0                 LISTENING       17196
  TCP    [::]:50095             [::]:0                 LISTENING       23396
  TCP    [fe80::a53d:3918:7a43:c312%11]:50710  [::]:0                 LISTENING       5032
  UDP    0.0.0.0:53             *:*                                    4556
  UDP    0.0.0.0:500            *:*                                    4620
  UDP    0.0.0.0:3389           *:*                                    1512
  UDP    0.0.0.0:4500           *:*                                    4620
  UDP    0.0.0.0:5050           *:*                                    10976
  UDP    0.0.0.0:5353           *:*                                    18224
  UDP    0.0.0.0:5353           *:*                                    18224
  UDP    0.0.0.0:5353           *:*                                    27548
  UDP    0.0.0.0:5353           *:*                                    27548
  UDP    0.0.0.0:5353           *:*                                    18224
  UDP    0.0.0.0:5353           *:*                                    27548
  UDP    0.0.0.0:5353           *:*                                    18224
  UDP    0.0.0.0:5353           *:*                                    3768
  UDP    0.0.0.0:5353           *:*                                    27548
  UDP    0.0.0.0:5355           *:*                                    3768
  UDP    0.0.0.0:6000           *:*                                    5184
  UDP    0.0.0.0:6881           *:*                                    26980
  UDP    0.0.0.0:36599          *:*                                    17340
  UDP    0.0.0.0:51039          *:*                                    17340
  UDP    0.0.0.0:51040          *:*                                    17340
  UDP    0.0.0.0:51041          *:*                                    17340
  UDP    0.0.0.0:51042          *:*                                    17340
  UDP    0.0.0.0:51043          *:*                                    17340
  UDP    0.0.0.0:51044          *:*                                    17340
  UDP    0.0.0.0:51045          *:*                                    17340
  UDP    0.0.0.0:51046          *:*                                    17340
  UDP    0.0.0.0:51047          *:*                                    17340
  UDP    0.0.0.0:51048          *:*                                    17340
  UDP    0.0.0.0:54723          *:*                                    4556
  UDP    0.0.0.0:54724          *:*                                    4556
  UDP    0.0.0.0:64071          *:*                                    27548
  UDP    0.0.0.0:64072          *:*                                    27548
  UDP    10.30.247.227:137      *:*                                    4
  UDP    10.30.247.227:138      *:*                                    4
  UDP    10.30.247.227:1900     *:*                                    3960
  UDP    10.30.247.227:55556    *:*                                    3960
  UDP    127.0.0.1:1900         *:*                                    3960
  UDP    127.0.0.1:7890         *:*                                    12560
  UDP    127.0.0.1:49664        *:*                                    3912
  UDP    127.0.0.1:55557        *:*                                    3960
  UDP    172.27.224.1:67        *:*                                    4556
  UDP    172.27.224.1:68        *:*                                    4556
  UDP    172.27.224.1:137       *:*                                    4
  UDP    172.27.224.1:138       *:*                                    4
  UDP    172.27.224.1:1900      *:*                                    3960
  UDP    172.27.224.1:55558     *:*                                    3960
  UDP    [::]:500               *:*                                    4620
  UDP    [::]:3389              *:*                                    1512
  UDP    [::]:4500              *:*                                    4620
  UDP    [::]:5353              *:*                                    27548
  UDP    [::]:5353              *:*                                    18224
  UDP    [::]:5353              *:*                                    3768
  UDP    [::]:5353              *:*                                    18224
  UDP    [::]:5353              *:*                                    27548
  UDP    [::]:5355              *:*                                    3768
  UDP    [::]:6881              *:*                                    26980
  UDP    [::]:54725             *:*                                    4556
  UDP    [::]:64072             *:*                                    27548
  UDP    [::1]:1900             *:*                                    3960
  UDP    [::1]:55554            *:*                                    3960
  UDP    [fe80::7f07:5f1b:d32f:1eb3%29]:1900  *:*                                    3960
  UDP    [fe80::7f07:5f1b:d32f:1eb3%29]:55555  *:*                                    3960
  UDP    [fe80::a53d:3918:7a43:c312%11]:1900  *:*                                    3960
  UDP    [fe80::a53d:3918:7a43:c312%11]:55553  *:*                                    3960
",
}
`;
